### DISCUSSION TITLE
Extracting the network using porespy

### DISCUSSION THREAD

**Discussion by sara-madi** [PMEAL/porespy#346] [Q&A]
I generated the artificial image coded in the "snow_advanced" and extracted the network using PoreSpy. I then used pickle to save the extracted network python dictionary as follows:

> import pickle
> dict_file = open('temp.dict', 'wb')
> pickle.dump(net, dict_file)
> dict_file.close()

After, I used OpenPNM to load the network:

> prj = op.io.PoreSpy.load('temp')
> net = prj.network
> op.io.XDMF.save(network=net, filename="out")
> print(prj.geometries())
> geo = prj.geometries()["geo_01"]
> net_health = net.check_network_health()
> op.topotools.trim(network=net, pores=net_health["trim_pores"])
> air = op.phases.Air(network=net)
> phys_air = op.physics.Standard(network=net, phase=air, geometry=geo)
> perm = op.algorithms.StokesFlow(network=net)
> perm.setup(phase=air)

However, when I am assigning the boundary conditions, I am getting the following error:

> Traceback (most recent call last):
>   File "<stdin>", line 1, in <module>
>   File "c:\users\legion\desktop\sara\openpnm-dev\openpnm\core\Base.py", line 856, in pores
>     ind = self._get_indices(element='pore', labels=labels, mode=mode)
>   File "c:\users\legion\desktop\sara\openpnm-dev\openpnm\core\Base.py", line 747, in _get_indices
>     union = union + self[element+'.'+item.split('.')[-1]]
>   File "c:\users\legion\desktop\sara\openpnm-dev\openpnm\network\GenericNetwork.py", line 155, in __getitem__
>     vals = super().__getitem__(key)
>   File "c:\users\legion\desktop\sara\openpnm-dev\openpnm\core\Base.py", line 272, in __getitem__
>     raise KeyError(key)

I think the problem is that openpnm can't find the pores(left, right,..). How can I solve this issue?

**Comment by jgostick** [comment]
You try to do ``print(net)`` and see what labels are available on your extracted network.  You can add your own boundary pores using [this example](https://github.com/PMEAL/OpenPNM/blob/master/examples/topology/Adding%20Boundary%20Pores.ipynb)

**Comment by sara-madi** [comment]
My labels are as follows:

>  Labels                                        Assigned Locations
――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――
1     pore.all                                      668
2     throat.all                                    3268
――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――
So, I should divide the total pores into other labels defined as: pore.left, pore.right, pore.bottom, pore.top, pore.front, pore.back. 
I don't think the example you provided(using clone) will allow me to create these new labels.

**Comment by jgostick** [comment]
You can have labels applied during the extraction process...I think they should be applied by default actually, so I'm surprized they're not here.  You can also add them to the network using spatial info, so:

``` python
Ps = net['pore.coords'][:, 0] < lower_lim
net.set_label(pores=Ps, label='pore.left')
Ps = net['pore.coords'][:, 0] > upper_lim
net.set_label(pores=Ps, label='pore.right')
```

Where upper_lim and lower_lim are physical distances that you consider to be 'on the right or left face' of the domain.  For instance:

``` python
lower_lim = pn['pore.coords'][:, 0].min() * 1.1
upper_lim = pn['pore.coords'][:, 0].min() * 0.9`
```

**Comment by sara-madi** [comment]
The "Snow_Advanced" extracted network does not give the labels of the pores; however,  the older version of snow algorithm given the following example
> https://github.com/PMEAL/OpenPNM/blob/master/examples/extractions/Working%20with%20Extracted%20Networks.ipynb

extracts the network specifying the labels of the pores as left, right,...
